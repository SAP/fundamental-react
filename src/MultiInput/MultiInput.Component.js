import { MultiInput } from '../';
import { Description, DocsText, DocsTile, Header, Import, Properties, Separator } from '../_playground';
import React, { Component } from 'react';

export class MultiInputComponent extends Component {
  data = [
      'Apple',
      'Apricot',
      'Acai',
      'African Moringa',
      'Bearberry',
      'Bilberry',
      'Blood orange',
      'Barbadine',
      'Barbados cherry',
      'Balsam Apple',
      'Chokeberry',
      'Cranberry',
      'Cupuacu'
  ];

  constructor(props) {
      super(props);

      this.state = {
          data: []
      };
  }

  performTagsUpdate = aTags => {
      alert(aTags);
  };

  multiInputCode = `<MultiInput
  data={this.data}
  onTagsUpdate={this.performTagsUpdate}
  placeHolder="Select a Fruit"
/>`;

  multiInputCompactCode = `<MultiInput
  data={this.data}
  onTagsUpdate={this.performTagsUpdate}
  placeHolder="Select a Fruit"
  compact={true}
/>`;

  render() {
      return (
          <div>
              <Header>Multi Input</Header>
              <Description />
              <Import module='MultiInput' path='/fundamental-react/src/' />

              <Separator />

              <Properties
                  properties={[
                      {
                          name: 'data',
                          description:
                'array (Required) - Collection of items to display in the list.'
                      },
                      {
                          name: 'placeHolder',
                          description:
                'string - The text to use as placeholder when no text is entered.'
                      },
                      {
                          name: 'onTagsUpdate',
                          description:
                'func - Method to fire on add or remove of tag. Component returns array of tags selected.'
                      },
                      {
                          name: 'compact',
                          description:
                'bool - true: display compact style, false: default style'
                      }
                  ]}
                  type='Inputs' />

              <Separator />

              <h2>Default</h2>
              <Description>
          A text input when on focus will show list of items to select.
              </Description>
              <DocsTile>
                  <div>
                      <MultiInput
                          data={this.data}
                          onTagsUpdate={this.performTagsUpdate}
                          placeHolder='Select a Fruit' />
                  </div>
              </DocsTile>
              <DocsText>{this.multiInputCode}</DocsText>

              <Separator />

              <h2>Compact Style</h2>
              <Description>
          A text input when on focus will show list of items to select, but with
          a compact input box.
              </Description>
              <DocsTile>
                  <div>
                      <MultiInput
                          compact
                          data={this.data}
                          onTagsUpdate={this.performTagsUpdate}
                          placeHolder='Select a Fruit' />
                  </div>
              </DocsTile>
              <DocsText>{this.multiInputCompactCode}</DocsText>

              <Separator />
          </div>
      );
  }
}
