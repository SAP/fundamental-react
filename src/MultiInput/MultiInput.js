import Button from '../Button/Button';
import Checkbox from '../Forms/Checkbox';
import classnamesBind from 'classnames/bind';
import { FORM_MESSAGE_TYPES } from '../utils/constants';
import FormInput from '../Forms/FormInput';
import FormMessage from '../Forms/_FormMessage';
import InputGroup from '../InputGroup/InputGroup';
import keycode from 'keycode';
import List from '../List/List';
import Popover from '../Popover/Popover';
import PropTypes from 'prop-types';
import requiredIf from 'react-required-if';
import shortid from '../utils/shortId';
import Token from '../Token/Token';
import withStyles from '../utils/withStyles';
import React, { Component } from 'react';
import checkboxStyles from 'fundamental-styles/dist/checkbox.css';
import inputGroupStyles from 'fundamental-styles/dist/input-group.css';
import listStyles from 'fundamental-styles/dist/list.css';
import tokenizerStyles from 'fundamental-styles/dist/tokenizer.css';

const classnames = classnamesBind.bind({
    ...checkboxStyles,
    ...inputGroupStyles,
    ...listStyles,
    ...tokenizerStyles
});
const isUsingCssModules = tokenizerStyles && Object.keys(tokenizerStyles).length > 0;

/** A **MultiInput** allows users to enter multiple values which are displayed as a tokens. It provides an editable input field for filtering the list,
 * and a dropdown menu with a list of the available options.
 * If the entries are not validated by the application, users can also enter custom values. */

class MultiInput extends Component {
    constructor(props) {
        super(props);

        this.state = {
            bShowList: false,
            tags: []
        };

        this.multiInputId = shortid.generate();
    }

    // create tags to display in dropdown list
    createTagList = data => {
        const {
            cssNamespace
        } = this.props;

        return data.map((item, index) => (
            <List.Item key={index}>
                <Checkbox
                    checked={this.isChecked(item)}
                    className={classnames(`${cssNamespace}-list__input`)}
                    compact={this.props.compact}
                    id={index + `_${this.multiInputId}`}
                    labelClassName={classnames(`${cssNamespace}-list__label`)}
                    onChange={() => this.updateSelectedTags(item)}
                    value={item}>
                    <List.Text className={classnames(`${cssNamespace}-checkbox__text`)}>{item}</List.Text>
                </Checkbox>
            </List.Item>
        ));
    };

    handleTagKeyDown = (event, tag) => {
        if (keycode(event) === 'backspace' || keycode(event) === 'delete') {
            this.removeTag(tag);
        }
    };

    // create tag elements to display below input box
    createTags = () => {
        const {
            cssNamespace
        } = this.props;

        return this.state.tags.map((tag, index) => {
            if (index < 3) {
                return (
                    <Token
                        className={isUsingCssModules && classnames(`${cssNamespace}-token`)}
                        key={index}
                        onClick={() => this.removeTag(tag)}
                        onKeyDown={(event) => this.handleTagKeyDown(event, tag)}>
                        {tag}
                    </Token>
                );
            } else if (index >= this.state.tags.length - 1) {
                return (
                    <span className={classnames(`${cssNamespace}-tokenizer__indicator`)}>{this.state.tags.length - 3} more</span>
                );
            } else {
                return null;
            }
        });
    };

    // add/remove tag to tag collection
    updateSelectedTags = (tag) => {

        if (this.state.tags.indexOf(tag) === -1) {
            this.setState(
                prevState => {
                    const tags = prevState.tags;
                    tags.push(tag);

                    return { tags: tags };
                },
                () => this.props.onTagsUpdate(this.state.tags)
            );
        } else {
            this.setState(
                prevState => {
                    let tags = prevState.tags.filter(item => {
                        return item.toLowerCase() !== tag.toLowerCase();
                    });

                    return { tags: tags };
                },
                () => this.props.onTagsUpdate(this.state.tags)
            );
        }
    };

    // check to see if tag is should be checked in list
    isChecked = tag => {
        if (this.state.tags.indexOf(tag) === -1) {
            return false;
        } else {
            return true;
        }
    };

    // remove/close tag
    removeTag = (tag) => {
        this.setState(
            prevState => {
                const tags = prevState.tags.filter(item => {
                    return item.toLowerCase() !== tag.toLowerCase();
                });

                return { tags: tags };
            },
            () => this.props.onTagsUpdate(this.state.tags)
        );
    };

    // show/hide tag list drop down
    showHideTagList = () => {
        this.setState(prevState => {
            return { bShowList: !prevState.bShowList };
        });
    };

    handleClickOutside = () => {
        this.setState({ bShowList: false });
    }

    render() {
        const {
            popoverProps,
            buttonLabel,
            buttonProps,
            compact,
            className,
            cssNamespace,
            disabled,
            data,
            listProps,
            inputProps,
            onTagsUpdate,
            placeholder,
            tagProps,
            validationOverlayProps,
            validationState,
            ...rest
        } = this.props;


        const tokenizerClassName = classnames(
            `${cssNamespace}-tokenizer`,
            {
                [`${cssNamespace}-tokenizer--compact`]: compact
            }
        );

        const listClassName = classnames(
            `${cssNamespace}-list--dropdown`,
            `${cssNamespace}-list--multi-input`,
            {
                [`${cssNamespace}-list--has-message`]: validationState?.state
            }
        );

        const inputGroupClasses = classnames(
            `${cssNamespace}-input-group--control`,
            {
                'is-disabled': disabled,
                [`is-${validationState?.state}`]: validationState?.state
            }
        );

        const popoverBody = (
            <List
                className={listClassName}
                compact={compact}
                {...listProps}>
                {this.createTagList(data)}
            </List>
        );

        const inputGroup = (
            <InputGroup
                {...rest}
                aria-expanded={this.state.bShowList}
                aria-haspopup='true'
                className={inputGroupClasses}
                compact={compact}
                disabled={disabled}
                onClick={this.showHideTagList}
                validationOverlayProps={validationOverlayProps}
                validationState={this.state.bShowList ? null : validationState}>
                <div {...tagProps} className={tokenizerClassName}>
                    <div className={classnames(`${cssNamespace}-tokenizer__inner`)}>
                        {this.state.tags.length > 0 && this.createTags()}
                        <FormInput
                            {...inputProps}
                            className={classnames(`${cssNamespace}-input-group__input`, `${cssNamespace}-tokenizer__input`, { [`${cssNamespace}-input`]: isUsingCssModules })}
                            compact={compact}
                            disabled={disabled}
                            placeholder={placeholder} />
                    </div>
                </div>
                <InputGroup.Addon isButton>
                    <Button
                        aria-label={buttonLabel}
                        {...buttonProps}
                        disabled={disabled}
                        glyph='value-help'
                        option='transparent' />
                </InputGroup.Addon>
            </InputGroup>
        );

        return (
            <Popover
                {...popoverProps}
                body={
                    (<>
                        {validationState &&
                        <FormMessage
                            {...validationOverlayProps?.formMessageProps}
                            forPopoverList
                            type={validationState.state}>
                            {validationState.text}
                        </FormMessage>
                        }
                        {popoverBody}
                    </>)}
                control={inputGroup}
                disableKeyPressHandler
                disabled={disabled}
                noArrow
                onClickOutside={this.handleClickOutside}
                useArrowKeyNavigation
                widthSizingType='matchTarget' />
        );
    }
}

MultiInput.displayName = 'MultiInput';

MultiInput.propTypes = {
    /** Collection of items to display in the list */
    data: PropTypes.array.isRequired,
    /** Localized string label for dropdown button's aria-label. This is required if `buttonProps` doesn't have a valid `aria-label`. */
    buttonLabel: requiredIf(PropTypes.string, props => !props?.buttonProps || !props.buttonProps['aria-label']?.trim()),
    /** Additional props to be spread to the `<button>` element */
    buttonProps: PropTypes.object,
    /** CSS class(es) to add to the element */
    className: PropTypes.string,
    /** Set to **true** to enable compact mode */
    compact: PropTypes.bool,
    /** Set to **true** to mark component as disabled and make it non-interactive */
    disabled: PropTypes.bool,
    /** Additional props to be spread to the `<input>` element */
    inputProps: PropTypes.object,
    /** Additional props to be spread to the `<ul>` element */
    listProps: PropTypes.object,
    /** Localized placeholder text of the input */
    placeholder: PropTypes.string,
    /** Additional props to be spread to the Popover component */
    popoverProps: PropTypes.object,
    /** Additional props to be spread to the tags `<div>` element */
    tagProps: PropTypes.object,
    /** Additional props to be spread to the ValidationOverlay */
    validationOverlayProps: PropTypes.shape({
        /** Additional classes to apply to validation popover's outermost `<div>` element  */
        className: PropTypes.string,
        /** Additional props to be spread to the ValdiationOverlay's FormMessage component */
        formMessageProps: PropTypes.object,
        /** Additional classes to apply to validation popover's popper child `<div>` wrapping the provided children  */
        innerRefClassName: PropTypes.string,
        /** Additional classes to apply to validation popover's popper `<div>` element  */
        popperClassName: PropTypes.string,
        /** CSS class(es) to add to the ValidationOverlay's reference `<div>` element */
        referenceClassName: PropTypes.string,
        /** Additional props to be spread to the popover's outermost `<div>` element */
        wrapperProps: PropTypes.object
    }),
    /** An object identifying a validation message.  The object will include properties for `state` and `text`; _e.g._, \`{ state: \'warning\', text: \'This is your last warning\' }\` */
    validationState: PropTypes.shape({
        /** State of validation: 'error', 'warning', 'information', 'success' */
        state: PropTypes.oneOf(FORM_MESSAGE_TYPES),
        /** Text of the validation message */
        text: PropTypes.string
    }),
    /**
     * Callback function; triggered when a tag is added or removed.
     *
     * @param {string[]} selectedTags array of selected tags as strings
     * @returns {void}
     * */
    onTagsUpdate: PropTypes.func
};

MultiInput.defaultProps = {
    onTagsUpdate: () => {}
};

export default withStyles(MultiInput);
