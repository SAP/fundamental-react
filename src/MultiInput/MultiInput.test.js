import { mount } from 'enzyme';
import MultiInput from './MultiInput';
import React from 'react';

describe('<MultiInput />', () => {
    const mockOnTagsUpdate = jest.fn();
    const data = [
        'Apple',
        'Apricot',
        'Acai',
        'African Moringa',
        'Bearberry',
        'Bilberry',
        'Blood orange',
        'Barbadine',
        'Barbados cherry',
        'Balsam Apple',
        'Chokeberry',
        'Cranberry',
        'Cupuacu'
    ];
    const getMultiInput = (props) => (
        <MultiInput
            {...props}
            data={data}
            onTagsUpdate={mockOnTagsUpdate}
            placeholder='Select a Fruit' />
    );
    const multiInput = getMultiInput();

    const compactMultiInput = (
        <MultiInput
            className='blue'
            compact
            data={data}
            onTagsUpdate={mockOnTagsUpdate}
            placeholder='Select a Fruit' />
    );

    let wrapper;

    const getListStatus = () => {
        const combobox = wrapper.find('.fd-input-group');

        const popover = wrapper.find(
            'div.fd-popover__popper'
        );

        return { combobox: combobox, popover: popover };
    };

    // create a default multi-input control
    test('create multi-input', () => {
        wrapper = mount(multiInput);

        expect(wrapper.find('.fd-list--multi-input')).toBeDefined();
    });

    // create a compact multi-input control
    test('create compact multi-input', () => {
        wrapper = mount(compactMultiInput);

        expect(wrapper.find('fd.input--compact')).toBeDefined();
    });

    // check that the tag list is hidden
    test('check that tag list is hidden', () => {
        wrapper = mount(multiInput);

        // check if bShowList state is changed
        expect(wrapper.children().state(['bShowList'])).toBe(false);

        // check to see if list is not shown
        let results = getListStatus(false);
        expect(results.combobox).toHaveLength(1);
        expect(results.popover).toHaveLength(0);
    });

    // check to display tag list on input text click
    test('check that tag list is shown when input text is clicked', () => {
        // check if bShowList state is changed
        wrapper = mount(multiInput);

        // simulate click on input text box
        wrapper.find('input[type="text"].fd-input').simulate('click');

        // check if bShowList state is changed
        expect(wrapper.children().state(['bShowList'])).toBe(true);

        // check to see if list is shown
        let results = getListStatus();
        expect(results.combobox).toHaveLength(1);
        expect(results.popover).toHaveLength(1);
    });

    test('check that multiple validation overlays are not displayed', () => {
        wrapper = mount(getMultiInput({
            validationState: { state: 'error', text: 'Test validation state' }
        }));

        wrapper.find('input[type="text"].fd-input').simulate('click');
        wrapper.find('input[type="text"].fd-input').simulate('focus');
        expect(document.querySelectorAll('.fd-list__message').length).toBe(1);

    });

    // check that tag list is shown on drop down click
    test('check that tag list is shown when dropdown button is clicked', () => {
        wrapper = mount(multiInput);

        // simulate click on dropdown button
        wrapper
            .find('button.fd-button--transparent')
            .simulate('click');

        // check if bShowList state is changed
        expect(wrapper.children().state(['bShowList'])).toBe(true);

        // check to see if list is shown
        let results = getListStatus();
        expect(results.combobox).toHaveLength(1);
        expect(results.popover).toHaveLength(1);
    });

    test('add tag to tagList', () => {
        wrapper = mount(multiInput);
        wrapper
            .find('button.fd-button--transparent')
            .simulate('click');

        // check that no tags exist
        expect(wrapper.children().state(['tags'])).toHaveLength(0);

        // add tag to list
        wrapper
            .find('.fd-checkbox').first()
            .simulate('change', { target: { value: data[0] } });

        // check that tag list contains value
        expect(wrapper.children().state(['tags'])).toHaveLength(1);

        // check to see if tag button is created
        expect(wrapper.find('span.fd-token[role="button"]')).toHaveLength(1);

        // check that tag text is correct
        expect(wrapper.find('span.fd-token[role="button"]').text()).toEqual(
            data[0]
        );
    });

    test('tag list is truncated when more than 3 options are selected', () => {
        wrapper = mount(multiInput);
        wrapper
            .find('button.fd-button--transparent')
            .simulate('click');

        // add 4 tags to list
        wrapper
            .find('.fd-checkbox').first()
            .simulate('change', { target: { value: data[0] } });
        wrapper
            .find('.fd-checkbox').at(1)
            .simulate('change', { target: { value: data[0] } });
        wrapper
            .find('.fd-checkbox').at(2)
            .simulate('change', { target: { value: data[0] } });
        wrapper
            .find('.fd-checkbox').at(3)
            .simulate('change', { target: { value: data[0] } });

        // check that tag list contains value
        expect(wrapper.children().state(['tags'])).toHaveLength(4);

        // check to see that only 3 tags are created
        expect(wrapper.find('span.fd-token[role="button"]')).toHaveLength(3);

        // check that truncated text is correct
        expect(wrapper.find('span.fd-tokenizer__indicator').text()).toEqual(
            '1 more'
        );
    });

    test('remove tag from taglist by unchecking', () => {
        wrapper = mount(multiInput);
        wrapper
            .find('button.fd-button--transparent')
            .simulate('click');

        // check that no tags exist
        expect(wrapper.children().state(['tags'])).toHaveLength(0);

        // add tag to list
        wrapper
            .find('.fd-checkbox').first()
            .simulate('change', { target: { value: data[0] } });

        // check that tag list contains value
        expect(wrapper.children().state(['tags'])).toHaveLength(1);

        // simulate unchecking tag from tag list
        wrapper
            .find('.fd-checkbox').first()
            .simulate('change', { target: { value: data[0] } });

        // check that no tags exist
        expect(wrapper.children().state(['tags'])).toHaveLength(0);
    });

    test('remove tag from taglist by clicking on tag', () => {
        wrapper = mount(multiInput);
        wrapper.find('.fd-input').simulate('click');

        // check that no tags exist
        expect(wrapper.children().state(['tags'])).toHaveLength(0);

        // add tag to list
        wrapper
            .find('.fd-checkbox').first()
            .simulate('change', { target: { value: data[0] } });

        // add another tag to list
        wrapper
            .find('li')
            .at(3)
            .find('.fd-checkbox')
            .simulate('change', { target: { value: data[2] } });

        // check that tag list contains value
        expect(wrapper.children().state(['tags'])).toHaveLength(2);

        // simulate clicking on tag in tag collection under input box
        wrapper.find('.fd-token__close').first().simulate('click');

        // check that no tags exist
        expect(wrapper.children().state(['tags'])).toHaveLength(1);
    });

    describe('validationOverlayProps', () => {

        test('pass validationOverlayProps to InputGroup', () => {
            const element = mount(
                <MultiInput
                    data={data}
                    validationOverlayProps={{
                        className: 'foo'
                    }} />
            );

            expect(
                element.find('InputGroup').at(1).prop('validationOverlayProps')
            ).toMatchObject({
                className: 'foo'
            });
        });
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Input Group component', () => {
            const element = mount(<MultiInput data={data} data-sample='Sample' />);

            expect(
                element.find('.fd-input-group').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MultiInput component\'s input element', () => {
            const element = mount(<MultiInput data={data} inputProps={{ 'data-sample': 'Sample' }} />);

            expect(
                element.find('input').at(0).getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MultiInput component\'s button element', () => {
            const element = mount(<MultiInput buttonProps={{ 'data-sample': 'Sample' }} data={data} />);

            expect(
                element.find('button').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MultiInput component\'s ul element', () => {
            const element = mount(<MultiInput data={data} listProps={{ 'data-sample': 'Sample' }} />);

            element.find('input[type="text"].fd-input').simulate('click');
            expect(
                element.find('ul').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MultiInput component\'s tags div element', () => {
            const element = mount(<MultiInput
                data={data}
                onTagsUpdate={mockOnTagsUpdate}
                tagProps={{ 'data-sample': 'Sample' }} />);

            element.find('input[type="text"].fd-input').simulate('click');
            element
                .find('.fd-checkbox').first()
                .simulate('change', { target: { value: data[0] } });

            expect(
                element.find('.fd-tokenizer').getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });

        test('should allow props to be spread to the MultiInput component\'s Popover component', () => {
            const element = mount(<MultiInput data={data} popoverProps={{ 'data-sample': 'Sample' }} />);

            expect(
                element.find('div.fd-popover').first().getDOMNode().attributes['data-sample'].value
            ).toBe('Sample');
        });
    });
});
