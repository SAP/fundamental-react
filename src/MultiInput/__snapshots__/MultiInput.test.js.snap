// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiInput /> create compact multi-input 1`] = `
<div
  className="fd-multi-input"
>
  <div
    className="fd-multi-input-field"
  >
    <div
      className="fd-popover"
    >
      <div
        className="fd-popover__control"
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          aria-label="Image label"
          className="fd-combobox-control"
        >
          <div
            className="fd-input-group fd-input-group--after fd-input-group--compact"
          >
            <input
              className="fd-input fd-input--compact"
              onClick={[Function]}
              placeholder="Select a Fruit"
              type="text"
            />
            <span
              className="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
            >
              <button
                className="fd-button--light sap-icon--navigation-down-arrow"
                onClick={[Function]}
              />
            </span>
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="fd-popover__body fd-popover__body--no-arrow"
      >
        <nav
          className="fd-menu"
        >
          <ul
            className="fd-menu__list"
          >
            <li>
              <label
                className="fd-menu__item"
                htmlFor="0_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="0_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Apple"
                />
                Apple
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="1_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="1_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Apricot"
                />
                Apricot
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="2_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="2_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Acai"
                />
                Acai
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="3_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="3_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="African Moringa"
                />
                African Moringa
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="4_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="4_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Bearberry"
                />
                Bearberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="5_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="5_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Bilberry"
                />
                Bilberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="6_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="6_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Blood orange"
                />
                Blood orange
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="7_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="7_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Barbadine"
                />
                Barbadine
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="8_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="8_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Barbados cherry"
                />
                Barbados cherry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="9_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="9_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Balsam Apple"
                />
                Balsam Apple
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="10_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="10_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Chokeberry"
                />
                Chokeberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="11_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="11_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Cranberry"
                />
                Cranberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="12_771464"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="12_771464"
                  onChange={[Function]}
                  type="checkbox"
                  value="Cupuacu"
                />
                Cupuacu
              </label>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
  
</div>
`;

exports[`MultiInput /> create multi-input 1`] = `
<div
  className="fd-multi-input"
>
  <div
    className="fd-multi-input-field"
  >
    <div
      className="fd-popover"
    >
      <div
        className="fd-popover__control"
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          aria-label="Image label"
          className="fd-combobox-control"
        >
          <div
            className="fd-input-group fd-input-group--after"
          >
            <input
              className="fd-input"
              onClick={[Function]}
              placeholder="Select a Fruit"
              type="text"
            />
            <span
              className="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
            >
              <button
                className="fd-button--light sap-icon--navigation-down-arrow"
                onClick={[Function]}
              />
            </span>
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="fd-popover__body fd-popover__body--no-arrow"
      >
        <nav
          className="fd-menu"
        >
          <ul
            className="fd-menu__list"
          >
            <li>
              <label
                className="fd-menu__item"
                htmlFor="0_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="0_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Apple"
                />
                Apple
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="1_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="1_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Apricot"
                />
                Apricot
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="2_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="2_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Acai"
                />
                Acai
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="3_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="3_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="African Moringa"
                />
                African Moringa
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="4_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="4_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Bearberry"
                />
                Bearberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="5_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="5_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Bilberry"
                />
                Bilberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="6_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="6_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Blood orange"
                />
                Blood orange
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="7_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="7_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Barbadine"
                />
                Barbadine
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="8_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="8_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Barbados cherry"
                />
                Barbados cherry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="9_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="9_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Balsam Apple"
                />
                Balsam Apple
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="10_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="10_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Chokeberry"
                />
                Chokeberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="11_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="11_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Cranberry"
                />
                Cranberry
              </label>
            </li>
            <li>
              <label
                className="fd-menu__item"
                htmlFor="12_657818"
              >
                <input
                  checked={false}
                  className="fd-checkbox"
                  id="12_657818"
                  onChange={[Function]}
                  type="checkbox"
                  value="Cupuacu"
                />
                Cupuacu
              </label>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
  
</div>
`;
